### PR TITLE
pm: some minor cleanups

### PR DESCRIPTION
**Pull request by gmarull**
Some minor cleanups, including:

- Introduce `PM_STATE_COUNT` to determine number of states (avoids fragile macro `PM_STATES_LEN`)
- Split stats code into its source file to improve code readability, and improve its structure/format a bit
- Rename some source files to make their purpose more clear

### CODE REVIEW

**Review by teburd — APPROVED** [review]
I like it

File: include/pm/device.h [hunk 1]
```diff
@@ -151,6 +151,22 @@ struct pm_device {
...
+ * pointer array with the same size of the actual devices with PM enabled. This
+ * is used internally by the PM subsystem to keep track of suspended devices
+ * during system power transitions.
+ *
+ * @param dev_name Device name.
+ */
+#define Z_PM_DEVICE_DEFINE_SLOT(dev_name)				\
+	static const Z_DECL_ALIGN(struct device *)			\
+	_CONCAT(Z_PM_DEVICE_NAME(dev_name), slot) __used		\
+	__attribute__((__section__(".z_pm_device_slots")))
```
Comment (keith-zephyr) [inline include/pm/device.h:168]: Do you also need to change `pm_suspend_devices()`? That routine loops over all devices, regardless of whether the device enables PM or not.

I think it makes sense to split this commit into a separate PR as this is doing more than name changes like the other commits.

**Review by keith-zephyr — CHANGES_REQUESTED** [review]

File: include/pm/device.h (hunk 1 above)
Comment (gmarull) [reply include/pm/device.h:168]: Even though it iterates over all devices, only those with `pm != NULL` have a chance to use a slot, since others are ignored. I'll drop this commit and open a new PR, since I have more stuff to push in that area.

File: subsys/pm/pm_stats.c [hunk 2]
```diff
@@ -32,44 +25,48 @@ STATS_NAME(pm_cpu_stats, state_last_cycles)
...
+static int pm_stats_init(const struct device *dev)
 {
-	for (int i = 0; i < CONFIG_MP_NUM_CPUS; i++) {
-		for (int j = 0; j < PM_STATE_COUNT; j++) {
-			snprintk(pm_cpu_stat_names[i][j], PM_STAT_NAME_LEN,
+	ARG_UNUSED(dev);
+
+	for (uint8_t i = 0U; i < CONFIG_MP_NUM_CPUS; i++) {
+		for (uint8_t j = 0U; j < PM_STATE_COUNT; j++) {
+			snprintk(names[i][j], PM_STAT_NAME_LEN,
```
[the remark this answers is not in the document]
Comment (gmarull) [reply subsys/pm/pm_stats.c:41]: Good idea, let me see if I can do it

[the remark this answers is not in the document]
Comment (teburd) [reply subsys/pm/pm_stats.c:41]: I tried this before but ran into something... can't recall what now though

[the remark this answers is not in the document]
Comment (gmarull) [reply subsys/pm/pm_stats.c:41]: This is the best macrobatics I've come up with (UTIL_LISTIFY can't be nested) (note that naming is re-arranged so that state comes before cpu index):
```
#define NAME_INIT(state_idx) "pm_state_" STRINGIFY(state_idx)
#define NAMES_INIT(cpu_idx, _)						       \
	{ FOR_EACH(NAME_INIT, ("_cpu_" STRINGIFY(cpu_idx) "_stats",),	       \
		   0, 1, 2, 3, 4, 5, 6) "_cpu_" STRINGIFY(cpu_idx) "_stats" },
static const char *names[CONFIG_MP_NUM_CPUS][PM_STATE_COUNT] = {
	UTIL_LISTIFY(CONFIG_MP_NUM_CPUS, NAMES_INIT)
};
```

not sure if it's worth sacrificing readability for something that is usually enabled to debug/measure performance.

File: include/pm/device.h (hunk 1 above)
Comment (gmarull) [reply include/pm/device.h]: https://github.com/zephyrproject-rtos/zephyr/pull/40576

File: subsys/pm/pm_stats.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (keith-zephyr) [reply subsys/pm/pm_stats.c:41]: Thanks for trying.  I agree the readability is better with the for loops.

**Review by keith-zephyr — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
